src/relax/ir/expr.cc

```diff
@@ -473,6 +473,23 @@ Function::Function(Array<Var> params, Expr body, Optional<StructInfo> ret_struct
     ret_struct_info = body_sinfo;
   }
 
+  bool is_pure = [&]() -> bool {
```

Comment:
How would this behave in the recursive or mutually recursive case? That was the reason for not inferring purity in the first place. Detecting those cases and warning the user that they need to be explicitly annotated would be a reasonable approach

Reply:
Currently, the struct info for a `GlobalVar` is determined only by the forward declaration, and not by its body.  The default for this is determined here ([here](https://github.com/apache/tvm/blob/main/python/tvm/script/parser/relax/parser.py#L247)), where a function is pure unless explicitly annotated otherwise.

We have a similar problem with return values, where the inferred return type of a function may not omitted, resulting in incorrect struct inference in the calling scope.  I think the long-term solution to both is the same: To represent the lack of information while parsing, and to infer full information as a post-proc.